lib/rules/no-else-return.js

```diff
@@ -88,6 +95,43 @@ module.exports = {
                     }
 
                     const endToken = sourceCode.getLastToken(node);
+
+                    /**
+                     * Dig into `else` branch and receive all root level declarations.
+                     * Checks only if @var foo is block scope declaration, @not @var x
+                     * @example
+                     *   function foo (){
+                     *      if(baz){
+                     *        return 42
+                     *      } else {
+                     *        let foo=1;
+                     *        {
+                     *          const x=10;
+                     *        }
+                     *        return foo
+                     *      }
+                     *   }
+                     *
+                     */
+                     const blockScopeDeclaration = scope => {
+                         return scope.childScopes.some(childScope => {
+                             const {
+                                 start,
+                                 end
+                             } = childScope.block;
+
+                             const elseBranch =
+                                 start === startToken.start && end === endToken.end;
+
+                             return elseBranch ? childScope.variables.some(variable => BLOCK_SCOPED_DECLARATION.test(variable.defs[0].node.type)) : false
```

Comment:
I don't know exactly how ESLint's scope handling works, but it should work if you only check if `childScope.variables.length !== 0`. That's because function scoped declarations like `var` or `function` are hoisted to the containing function scope.

Reply:
I agree that it's worth trying out this simplification, although I think the real problem is that we need to track all variables available in the containing scope and look for declarations with the same name (which are shadowing before the autofix, but redeclaring after).